### PR TITLE
Consider only partitions in query for materialized view optimization

### PR DESCRIPTION
**Pull request by Sullivan-Patrick**
Test plan - Integration test covering direct view queries as well as automatic optimization.

This change makes materialized view optimizations consider the domain of the query when deciding if there are too many partitions missing to perform the optimization. For example, consider the following:

```SELECT a, b FROM tbl WHERE ds > '2022-03-09' GROUP BY a, b```

Assuming we had many stale partitions before `2022-03-09`, this query would fail to optimize using a backing materialized view, and increment a `TOO_MANY_PARTITIONS_MISSING` metric, even if every partition in the domain `> '2022-03-09'` was fresh. This PR changes that by only counting missing partitions inside of the filter's domain.

This PR also adds type coercion in `MaterializedViewQueryOptimizer` in the first commit. This is necessary to avoid type-mismatch related `IllegalArgumentException`s when building `RowExpression`s.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:298]: This does not seem right to me. Why do we need a root level field?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1445]: Very complex explanation.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1450]: Can we create the config for this feature (where we consider base query domain) we should be able to disable this if required.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1447]: Lets create a github issue for this. Good catch

Comment (Sullivan-Patrick) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1447]: Will do- I'd wait until this is merged, just so any details are unchanging for when I write the issue.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java
Comment (pranjalssh) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2455]: Previously we were checking for missing partitions per base table, now we are checking for their sum across all base tables. What is the reason behind this change?

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1189]: I think we should just enable it by default. @jainxrohit what do you think?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1476]: Don't catch all exceptions. Also we have this code in 2 places. Can we create a common function which does this try catch so the logic is isolated in one place

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (pranjalssh) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1446]: This is like reinventing predicate pushdown in analyzer

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java
Comment (Sullivan-Patrick) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2455]: I think if there are 0 missing partitions we should consider it as fully materialized- there is the possibility we hit the condition `!dataPredicates.getPredicateDisjuncts().isEmpty()` but filter any missing predicates now.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (jainxrohit) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:299]: I think we should fix this, let me think about it.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (jainxrohit) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1189]: Yes enabled by default. We need this config so that we can disable it if this runs in to any issues.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (jainxrohit) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:299]: Can we check if we can use Analysis object to store this information?

Comment (Sullivan-Patrick) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:299]: @jainxrohit I could add it directly as a field to `Analysis` which would be no problem. Did you have something more than that in mind?

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (Sullivan-Patrick) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1189]: Is it not enabled by default as written? `featuresConfig.isMarerializedViewDataConsistencyEnabled()` resolves to `true` from what I can tell.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (pranjalssh) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:299]: One way of doing it would be to rewrite all logic thats run when analyzeFrom is called again. We should only need some small parts which affect the scope

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1469]: [Major] Once we're here, the query is from a materialized view. So `baseQueryDomain` variable contains predicates on materialized view, and not the base table. We need to do some work to convert these predicates to that of base table

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestMaterializedViewQueryOptimizer.java
Comment (Sullivan-Patrick) [reply presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestMaterializedViewQueryOptimizer.java:200]: Yeah, this failed due to type mismatching. `d` is a `bigint`.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1200]: enable by default

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java:733]: These comments and newlines should be in previous commit

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (Sullivan-Patrick) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1200]: See previous: https://github.com/prestodb/presto/pull/18018#discussion_r924798708

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java:171]: Add a comment to explain this

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1487]: mappedToOneTable

**Review by pranjalssh — APPROVED** [review]

**Comment by pranjalssh** [comment]
Also, this change breaks presto-facebook. Please include a PR for that as well

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:2015]: The current config name is confusing. lets rename it to something else, how about?
"consider-query-filters-for-materialized-view-partitions"

**Review by jainxrohit — COMMENTED** [review]
overall looks in pretty good state

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java:172]: rename it to currentQuerySpecification

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1476]: Can this be moved to materializedViewUtils class?

Comment (Sullivan-Patrick) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1476]: Not trivially AFAICT- the `getMaterializedViewStatus` in `MaterializedViewQueryOptimizer` and the one in `StatementAnalyzer` do quite a few things differently. `StatementAnalyzer` for example requires getting mappings from the view/rewritten query to base table, while `MaterializedViewQueryOptimizer` has access to the base query and doesn't need to worry about them.

I could extract a couple of short helper methods, but not sure how much that would help readability.

**Review by pranjalssh — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java:776]: I am wondering if there is a need to rewrite if coercion information is not available? We should check that first before issuing defaultRewrite?

**Review by jainxrohit — COMMENTED** [review]
Check commit 2

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java
Comment (Sullivan-Patrick) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java:783]: _`analysis.getCoercion` directly hits a map, so AFAICT it is possible for this to be null, if `original` is not a key in the map. For example, if I run a unit test with view filter `WHERE a > CAST(5 AS bigint)`, this causes `coercion==null`.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:301]: do we still need it?

Comment (Sullivan-Patrick) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Not anymore, good catch.

**Review by jainxrohit — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java
Comment (Sullivan-Patrick) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java:775]: discussed offline

**Review by jainxrohit — APPROVED** [review]

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java
Comment (highker) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java:602]: A few comments here:
- Just realized that we use `String` in `MaterializedViewStatus` to represent a column. That probably is not ideal. We should use `ColumnHandle` and call them `columns` instead of `columnNames`. (Note that you can still compare between `ColumnHandle`s). If you have time, you can refactor the code to fix the issue in another PR.
- Please add java doc for this interface to explain the use of `baseQueryDomain`. That is important.

**Review by highker — APPROVED** [review]

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java
Comment (Sullivan-Patrick) [reply presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java:604]: Thanks for the feedback! I'll put the former in my to-do, seems like a pretty small ask I can do before I leave. If not I'll file an issue. Added the javadoc, let me know if you think it's descriptive.

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java
Comment (highker) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java:609]: Just realized this: why do we still keep this interface? Can we fix presto-facebook as well to remove it completely?

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/classloader/ClassLoaderSafeConnectorMetadata.java
Comment (highker) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/classloader/ClassLoaderSafeConnectorMetadata.java:521]: same here: otherwise this PR won't be triggered at all in prod env....

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java
Comment (highker) [reply presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java:609]: You can send out another PR to fix this BTW

**Pull request merged**
